### PR TITLE
Compatible version of typed Index And Get API

### PR DESCRIPTION
**Pull request by pgomulka**
Providing a REST infrastructure to handle compatible APIs. Rest APIs that were removed in 8 will be available only if an HTTP request had a compatible header. At the moment is is expecting an Accept header to contain `application/vnd.elasticsearch+json;compatible-with=7`. This might be changed though when https://github.com/elastic/elasticsearch/issues/52370 is resolved.
The REST changes contain
- enriching RestRequest with a parameter `Compatible-With`. The value is taken from `Accept` header. See RestRequest creation. Used when deciding if a request to a compatible handler can be dispatched. See RestController. 
//TODO this at the moment can be simplified to only use a value from a header. However in rest layer we often have only access to Params, so headers won't be accessible when implementing some compatible code

- enriching XContentBuilder with a compatible version value. See AbstractRestChannel. Used when shape of a response is changed. see DocWriteResponse and GetResult 

- MediaType parsing change - see XContentType. because we expect a different media type now `application/vnd.elasticsearch+json;compatible-with=7` (when previously application/json or similar)

- testing changes - compat tests will have a compatible header set. see AbstractCompatRestTest.java

- Compatible rest handlers - based on RestIndexAction and RestGetAction. See modules/rest-compatibility. These extend already existing handlers and register them in a RestCompatPlugin under a typed paths. They will only be accessible when a compatible header is present (use of compatibilityRequired method from RestHandler interface)

Without this PR 283 tests from 7.x are failing. - rest api spec tests only
with this PR 228 tests are passing - with almost all tests from index/* package passed (missing 2 tests that require change for include_type_param, to be added in a new PR)

relates #51816

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/rest/action/document/RestGetAction.java [hunk 1]
```diff
@@ -86,4 +100,30 @@ protected RestStatus getStatus(final GetResponse response) {
         });
     }
 
+    public static class CompatibleRestGetAction extends RestGetAction {
```
Comment (jakelandis) [inline server/src/main/java/org/elasticsearch/rest/action/document/RestGetAction.java]: Until we defined a better approach, can you start adding in something like this
```
assert Version.CURRENT.major == 8 : "REST API compatilbity for version 7 is only supported on version 8";```

File: server/src/main/java/org/elasticsearch/rest/CompatibleHandlers.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import org.elasticsearch.index.mapper.MapperService;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.UnaryOperator;
+
+public class CompatibleHandlers {
```
Comment (jakelandis) [inline server/src/main/java/org/elasticsearch/rest/CompatibleHandlers.java]: Can you add a TODO in here to ensure we get some buy in on https://github.com/elastic/elasticsearch/issues/52370 before we merge the header format.

**Comment by pgomulka** [comment]
ok to test

File: server/src/main/java/org/elasticsearch/rest/action/document/RestIndexAction.java [hunk 3]
```diff
@@ -19,26 +19,40 @@
...
 import static org.elasticsearch.rest.RestRequest.Method.PUT;
 
 public class RestIndexAction extends BaseRestHandler {
+    private static final String TYPES_DEPRECATION_MESSAGE = "[types removal] Specifying types in document " +
+        "index requests is deprecated, use the typeless endpoints instead (/{index}/_doc/{id}, /{index}/_doc, " +
+        "or /{index}/_create/{id}).";
+    private static final DeprecationLogger deprecationLogger = new DeprecationLogger(
+        LogManager.getLogger(RestIndexAction.class));
+    private static final Consumer<RestRequest> DEPRECATION_WARNING =
+        r -> deprecationLogger.deprecatedAndMaybeLog("index_with_types",TYPES_DEPRECATION_MESSAGE);
```
Comment (jakelandis) [inline server/src/main/java/org/elasticsearch/rest/action/document/RestIndexAction.java]: I think the above can be removed ?

File: server/src/main/java/org/elasticsearch/rest/CompatibleHandlers.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.function.Consumer;
+
+public class CompatibleHandlers {
+
+    /**
+     * TODO revisit when https://github.com/elastic/elasticsearch/issues/52370 is resolved
+     */
+    public static final String COMPATIBLE_HEADER = "Accept";
+    public static final String COMPATIBLE_PARAMS_KEY = "Compatible-With";
+    public static final String COMPATIBLE_VERSION = "7";
```
Comment (jakelandis) [inline server/src/main/java/org/elasticsearch/rest/CompatibleHandlers.java]: Can we make this major version - 1 , so that when v9 it is auto incremented to 8 ?

File: server/src/main/java/org/elasticsearch/rest/CompatibleHandlers.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    public static Consumer<RestRequest> consumeParameterType(DeprecationLogger deprecationLogger) {
```
Comment (jakelandis) [inline server/src/main/java/org/elasticsearch/rest/CompatibleHandlers.java]: this method and the one below feel like REST v7 concerns. Can they be moved to the module ?

(which would leave this class pretty empty and possibly move the constant strings elsewhere ?)

File: server/src/main/java/org/elasticsearch/rest/RestController.java [hunk 6]
```diff
@@ -328,7 +328,13 @@ private void tryAllHandlers(final RestRequest request, final RestChannel channel
                       return;
                   }
                 } else {
-                    dispatchRequest(request, channel, handler);
+                    if(handler.compatibilityRequired() == false //regular (not removed) handlers are always dispatched
+                        //handlers that were registered compatible, require request to be compatible
+                        || CompatibleHandlers.isV7Compatible(request)) {
```
Comment (jakelandis) [inline server/src/main/java/org/elasticsearch/rest/RestController.java]: any way to remove the v7 here ? ideally this would be agnostic to the actual version (which I think it already is in all but the name)

File: server/src/main/java/org/elasticsearch/rest/RestController.java [hunk 7]
```diff
@@ -340,6 +346,13 @@ private void tryAllHandlers(final RestRequest request, final RestChannel channel
         handleBadRequest(uri, requestMethod, channel);
     }
 
+    private void handleCompatibleNotAllowed(String rawPath, RestChannel channel) throws IOException {
+        String msg = "Compatible api can be only accessed with Compatible Header. Path used: " + rawPath;
```
Comment (jakelandis) [inline server/src/main/java/org/elasticsearch/rest/RestController.java]: I am on the fence about this message, but leaning towards removing it.  I am not sure we need to tell people that this could be a compatible path.

File: qa/rest-compat-tests/src/main/java/org/elasticsearch/rest/compat/AbstractCompatRestTest.java [hunk 8]
```diff
@@ -76,11 +77,17 @@ private static void mutateTestCandidate(ClientYamlTestCandidate testCandidate) {
             DoSection doSection = (DoSection) ds;
             //TODO: be more selective here
             doSection.setIgnoreWarnings(true);
-            //TODO: use the real header compatibility header
-            doSection.getApiCallSection().addHeaders(Collections.singletonMap("compatible-with", "v7"));
+
```
Comment (jakelandis) [inline qa/rest-compat-tests/src/main/java/org/elasticsearch/rest/compat/AbstractCompatRestTest.java]: note to other reviewers - the qa/rest-compat-tests is a temporary test fixture and currently runs the core REST rest tests from bwc:minor against a single current test cluster.  No need to review anything out of qa/rest-compat-tests since it will be going away in favor of per module/plugin tests. 

The (diminishing) number of failed core REST tests is what is guiding these early efforts.

File: server/src/test/java/org/elasticsearch/common/xcontent/XContentTypeTests.java [hunk 9]
```diff
@@ -84,4 +84,11 @@ public void testFromRubbish() throws Exception {
         assertThat(XContentType.fromMediaTypeOrFormat("text/plain"), nullValue());
         assertThat(XContentType.fromMediaTypeOrFormat("gobbly;goop"), nullValue());
     }
+
+    public void testMediaType() throws Exception {
+        String mediaType = XContentType.parseMediaType("application/vnd.elasticsearch+json;compatible-with=7");
+        assertThat(mediaType,equalTo("application/json"));
+        mediaType = XContentType.parseMediaType("application/json");
+        assertThat(mediaType,equalTo("application/json"));
+    }
```
Comment (jakelandis) [inline server/src/test/java/org/elasticsearch/common/xcontent/XContentTypeTests.java]: can you also test yaml,cbor, and smile ?

also, does case matter ? (can you add a test for that too)

File: server/src/test/java/org/elasticsearch/common/xcontent/XContentTypeTests.java [hunk 10]
```diff
@@ -84,4 +84,11 @@ public void testFromRubbish() throws Exception {
         assertThat(XContentType.fromMediaTypeOrFormat("text/plain"), nullValue());
         assertThat(XContentType.fromMediaTypeOrFormat("gobbly;goop"), nullValue());
     }
+
+    public void testMediaType() throws Exception {
+        String mediaType = XContentType.parseMediaType("application/vnd.elasticsearch+json;compatible-with=7");
```
Comment (jakelandis) [inline server/src/test/java/org/elasticsearch/common/xcontent/XContentTypeTests.java]: can you use a random value for the compatible with (to help catch any regex assumptions around single digits)

File: server/src/test/java/org/elasticsearch/common/xcontent/XContentTypeTests.java [hunk 11]
```diff
@@ -84,4 +84,11 @@ public void testFromRubbish() throws Exception {
...
+        assertThat(mediaType,equalTo("application/json"));
+        mediaType = XContentType.parseMediaType("application/json");
+        assertThat(mediaType,equalTo("application/json"));
```
Comment (jakelandis) [inline server/src/test/java/org/elasticsearch/common/xcontent/XContentTypeTests.java]: can you also test parseVersion ?

File: modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/RestCompatPlugin.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    public List<RestHandler> getRestHandlers(
+        Settings settings,
+        RestController restController,
+        ClusterSettings clusterSettings,
+        IndexScopedSettings indexScopedSettings,
+        SettingsFilter settingsFilter,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        Supplier<DiscoveryNodes> nodesInCluster
+    ) {
+        return List.of(
```
Comment (jakelandis) [inline modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/RestCompatPlugin.java]: I think can wrap this registration in an "if version == 8", so when v9 is cut, the code would effectively be dead.

File: server/src/test/java/org/elasticsearch/action/DocWriteResponseTests.java [hunk 13]
```diff
@@ -107,4 +107,34 @@ public void testToXContentDoesntIncludeForcedRefreshUnlessForced() throws IOExce
             }
         }
     }
+
+    public void testTypeWhenCompatible() throws IOException {
```
Comment (jakelandis) [inline server/src/test/java/org/elasticsearch/action/DocWriteResponseTests.java]: Is it possible to move this test (and any other compatible tests) over to the compatible module under the version7 package ?  Ideally, we keep this coverage, but make it super easy to remove.

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/MatchAssertion.java [hunk 14]
```diff
@@ -56,6 +56,12 @@ public MatchAssertion(XContentLocation location, String field, Object expectedVa
 
     @Override
     protected void doAssert(Object actualValue, Object expectedValue) {
+        // TODO this needs to be moved to override directory
```
Comment (jakelandis) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/MatchAssertion.java]: ++ this needs to be handled by the compat test frame work, but is good here for now.

File: server/src/test/java/org/elasticsearch/http/DefaultRestChannelTests.java [hunk 15]
```diff
@@ -212,6 +214,14 @@ public void testHeadersSet() {
         assertEquals(resp.contentType(), headers.get(DefaultRestChannel.CONTENT_TYPE).get(0));
     }
 
+    public void testCompatibleParamIsSet(){
+        final TestRequest httpRequest = new TestRequest(HttpRequest.HttpVersion.HTTP_1_1, RestRequest.Method.GET, "/");
+        httpRequest.getHeaders().put(HttpHeaders.ACCEPT, List.of("application/vnd.elasticsearch+json;compatible-with=7"));
```
Comment (jakelandis) [inline server/src/test/java/org/elasticsearch/http/DefaultRestChannelTests.java]: can you use a random version ?

File: server/src/test/java/org/elasticsearch/rest/RestControllerTests.java [hunk 16]
```diff
@@ -623,6 +623,39 @@ public void testDispatchRestrictSystemIndices() {
         assertFalse(context.isSystemIndexAccessAllowed());
     }
 
+    public void testDispatchCompatibleHandler() {
+        final String mimeType = randomFrom("application/vnd.elasticsearch+json;compatible-with=7");
```
Comment (jakelandis) [inline server/src/test/java/org/elasticsearch/rest/RestControllerTests.java]: can you use a random version ?

File: server/src/test/java/org/elasticsearch/rest/RestControllerTests.java [hunk 17]
```diff
@@ -623,6 +623,39 @@ public void testDispatchRestrictSystemIndices() {
...
+        String content = randomAlphaOfLength((int) Math.round(BREAKER_LIMIT.getBytes() / inFlightRequestsBreaker.getOverhead()));
+        final List<String> contentTypeHeader = Collections.singletonList(mimeType);
+        FakeRestRequest fakeRestRequest = new FakeRestRequest.Builder(NamedXContentRegistry.EMPTY)
+            .withContent(new BytesArray(content), RestRequest.parseContentType(contentTypeHeader)).withPath("/foo")
+            .withHeaders(Map.of("Content-Type", contentTypeHeader, "Accept", contentTypeHeader))
```
Comment (jakelandis) [inline server/src/test/java/org/elasticsearch/rest/RestControllerTests.java]: Here we are passing in both Content-Type and Accept headers. Should we test one or the other or both ? (if so a random selection would be good)

File: server/src/test/java/org/elasticsearch/rest/RestControllerTests.java [hunk 18]
```diff
@@ -623,6 +623,39 @@ public void testDispatchRestrictSystemIndices() {
...
+            .build();
```
Comment (jakelandis) [inline server/src/test/java/org/elasticsearch/rest/RestControllerTests.java]: Also, what happens if you set the Content-Type to application/json, but Accept to application/vnd.elasticsearch+json;compatible-with=7 ?

I wonder if we need more testing for these edge cases ?(either unit or integration)

File: server/src/main/java/org/elasticsearch/rest/RestController.java (hunk 6 above)
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/rest/RestController.java]: good point, the generic and always present compatible code can be made agnostic.

File: server/src/main/java/org/elasticsearch/rest/CompatibleHandlers.java (hunk 4 above)
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/rest/CompatibleHandlers.java]: good idea, this one will always stay as it is part of generic code always used for compatible API

File: server/src/main/java/org/elasticsearch/rest/CompatibleHandlers.java (hunk 5 above)
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/rest/CompatibleHandlers.java]: good point, the isV7Compatible can be made generic (applying to all compatible versions in the future) and can stay in RestController. 
How about the constants though? any ideas where to place them? It feels to me like `org.elasticsearch.rest` package would be good. Maybe RestUtils?

File: server/src/main/java/org/elasticsearch/rest/RestController.java (hunk 7 above)
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/rest/RestController.java]: I guess for future clients versions there would always be passed a header with a version. But in transition from 7 to 8, I guess some people would stay with old (like 7.0) client and won't have a header? This way they could find out quickly about the compatible API.

I added it mostly because I found it useful when developing - but happy to remove it too.

File: modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/RestCompatPlugin.java (hunk 12 above)
Comment (pgomulka) [reply modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/RestCompatPlugin.java]: great idea, it will also be useful in a transition period when some new compatible apis (aiming 9) are added before the old code is removed

File: server/src/test/java/org/elasticsearch/rest/RestControllerTests.java (hunk 18 above)
Comment (pgomulka) [reply server/src/test/java/org/elasticsearch/rest/RestControllerTests.java]: good question. we can be more specific for response generation with Accept and request parsing with Content-Type.
But on the other hand I think it will just complicate matters too much. 
I would prefer to assume (and assert in code) that these are the same

File: server/src/test/java/org/elasticsearch/rest/RestControllerTests.java (hunk 17 above)
Comment (pgomulka) [reply server/src/test/java/org/elasticsearch/rest/RestControllerTests.java]: as per https://github.com/elastic/elasticsearch/pull/53228#discussion_r391180225
I think we should assume both should be present.
WDYT? 
I will then reply with the conclusion on https://github.com/elastic/elasticsearch/issues/52370

File: server/src/test/java/org/elasticsearch/action/DocWriteResponseTests.java (hunk 13 above)
Comment (pgomulka) [reply server/src/test/java/org/elasticsearch/action/DocWriteResponseTests.java]: good idea, I think the v7 subdirectory in rest-compatibility and tweaking gradle testClasses should be enough. 
so that we can have the same packages with same classes but in different submodules (the situation when v7 is not yet removed, but more compatible code for v8 is being created)
otherwise suffixing test classes with v7?

File: server/src/main/java/org/elasticsearch/rest/CompatibleHandlers.java (hunk 5 above)
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/rest/CompatibleHandlers.java]: `COMPATIBLE_HEADER` is only used in `RestRequest` (so there for now) (as a side note, we also need to also check `Content-Type`, but am good with that in a follow up)
`COMPATIBLE_VERSION` is only in `RestRequest` (so there for now)
`COMPATIBLE_PARAMS_KEY` is an x-content concept, so probably somewhere in x-content

File: server/src/main/java/org/elasticsearch/rest/RestController.java (hunk 7 above)
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/rest/RestController.java]: Let's remove for now. We can add it later if need be.

File: server/src/test/java/org/elasticsearch/rest/RestControllerTests.java (hunk 17 above)
Comment (jakelandis) [reply server/src/test/java/org/elasticsearch/rest/RestControllerTests.java]: Let's punt on it for now, the details there are important, but not immediately. I added a "parking lot" section on https://github.com/elastic/elasticsearch/issues/51816 so we don't forget.

File: server/src/test/java/org/elasticsearch/rest/RestControllerTests.java (hunk 18 above)
Comment (jakelandis) [reply server/src/test/java/org/elasticsearch/rest/RestControllerTests.java]: Same as above (I added a "parking lot" section on #51816 so we don't forget.)

File: server/src/test/java/org/elasticsearch/action/DocWriteResponseTests.java (hunk 13 above)
Comment (jakelandis) [reply server/src/test/java/org/elasticsearch/action/DocWriteResponseTests.java]: sounds good.
Since it mostly cosmetic, lets remove the v7 package, and only using the v7 in the class name. That should be sufficient to make it obvious it is v7 code and still allow us use package scope without creating new sub modules. We re-visit later if turns out too ugly :)

File: modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/version7/RestGetActionV7.java [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+
+import static org.elasticsearch.rest.RestRequest.Method.GET;
+import static org.elasticsearch.rest.RestRequest.Method.HEAD;
+
+public class RestGetActionV7 extends RestGetAction {
+
+    private static final DeprecationLogger deprecationLogger = new DeprecationLogger(LogManager.getLogger(RestGetAction.class));
+    private static final String TYPES_DEPRECATION_MESSAGE = "[types removal] Specifying types in "
+        + "document get requests is deprecated, use the /{index}/_doc/{id} endpoint instead.";
+    private static final Consumer<RestRequest> DEPRECATION_WARNING = r -> deprecationLogger.deprecatedAndMaybeLog(
```
Comment (jtibshirani) [inline modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/version7/RestGetActionV7.java]: Could we avoid introducing a `Consumer` here and just call `deprecationLogger.deprecatedAndMaybeLog(...)` directly inside `prepareRequest`?

File: modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/version7/CompatibleHandlers.java [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.util.function.Consumer;
+
+public class CompatibleHandlers {
+
+    public static Consumer<RestRequest> consumeParameterType(DeprecationLogger deprecationLogger) {
+        String TYPES_DEPRECATION_MESSAGE = "[types removal] Using type as a path parameter is deprecated.";
+
+        return r -> {
+            deprecationLogger.deprecatedAndMaybeLog("create_index_with_types", TYPES_DEPRECATION_MESSAGE);
```
Comment (jtibshirani) [inline modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/version7/CompatibleHandlers.java]: If I'm understanding correctly, it seems like we'll log two deprecation messages per request, both the endpoint-specific one like "Specifying types in document get requests is deprecated..." and this more general one "Using type as a path parameter is deprecated". To me we could just stick with the end-point specific one to avoid warning twice about the same issue.

As a side note, this general deprecation message is also logged under the key "create_index_with_types", I think we'd want a more precise name.

File: server/src/main/java/org/elasticsearch/action/DocWriteResponse.java [hunk 21]
```diff
@@ -276,6 +280,9 @@ public void writeTo(StreamOutput out) throws IOException {
     @Override
     public final XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject();
+        if (builder.getCompatibleMajorVersion() == Version.V_7_0_0.major) {
+            builder.field(TYPE_FIELD_NAME, SINGLE_MAPPING_TYPE);
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/action/DocWriteResponse.java]: I think you can just do the following, I've never had to wrap a string inside a `Text` object:

```
builder.field(TYPE_FIELD_NAME, MapperService.SINGLE_MAPPING_NAME);
```

File: server/src/test/java/org/elasticsearch/action/DocWriteResponseTests.java [hunk 22]
```diff
@@ -107,4 +107,6 @@ public void testToXContentDoesntIncludeForcedRefreshUnlessForced() throws IOExce
             }
         }
     }
+
```
Comment (jtibshirani) [inline server/src/test/java/org/elasticsearch/action/DocWriteResponseTests.java]: Small comment, would be good to revert this.

**Review by jtibshirani — COMMENTED** [review]
I added a couple comments from the types deprecation side. One high-level comment: we should ensure there's test coverage for all the deprecation messages we're adding back. In 7.x we had dedicated unit tests in the various actions -- [here's an example](https://github.com/elastic/elasticsearch/blob/7.6/server/src/test/java/org/elasticsearch/rest/action/document/RestIndexActionTests.java#L60) for `RestIndexAction`.

File: server/src/main/java/org/elasticsearch/rest/RestController.java (hunk 7 above)
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/rest/RestController.java]: will do. btw we are at the moment returning 400 bad request when handler was not found.

File: server/src/test/java/org/elasticsearch/action/DocWriteResponseTests.java (hunk 13 above)
Comment (pgomulka) [reply server/src/test/java/org/elasticsearch/action/DocWriteResponseTests.java]: I only was referring to tests - that if we have a copy of a class from a different module we could sufix with v7.
it feels right to have a version7 package in main though. as on the main `org.elasticsearch.rest.compat` level we could keep version agnostic classes - like RestCompatPlugin.

File: modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/version7/CompatibleHandlers.java [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+package org.elasticsearch.rest.compat.version7;
+
+import org.elasticsearch.common.logging.DeprecationLogger;
+import org.elasticsearch.rest.RestRequest;
+
+import java.util.function.Consumer;
+
+public class CompatibleHandlers {
+
+    public static Consumer<RestRequest> consumeParameterType(DeprecationLogger deprecationLogger) {
```
[the remark this answers is not in the document]
Comment (pgomulka) [reply modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/version7/CompatibleHandlers.java]: agree with both - method rename and making this a method. 
Originally the reason for this  being a consumer was because I was trying to fit in into the previous style of registering handlers. Extending the RestController.registerHandler with additional handlerWrappers. This is no longer needed after the recent refactoring made by Jay

File: modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/version7/RestGetActionV7.java (hunk 19 above)
Comment (pgomulka) [reply modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/version7/RestGetActionV7.java]: sure, that similarly to previous use of consumer was influenced by previous style of registering handlers.
a direct call is better here.

File: modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/version7/CompatibleHandlers.java (hunk 20 above)
Comment (pgomulka) [reply modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/version7/CompatibleHandlers.java]: agree - let's stick with just per endpoint warning.
WIth that I feel like we can even get rid of `CompatibleHandlers.consumeParameterType` - it would only consume a type param -  and the whole utility class. 
I was worried that it would end up a bag of all the helpers for compatibility. It might generate a little bit of duplication, but there is no need for it now.

Do you mean more precise name for a deprecation key? Like to be the same for Get and Index and other handlers?

File: server/src/main/java/org/elasticsearch/action/DocWriteResponse.java (hunk 21 above)
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/action/DocWriteResponse.java]: will do

File: modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/RestCompatPlugin.java [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+        RestController restController,
+        ClusterSettings clusterSettings,
+        IndexScopedSettings indexScopedSettings,
+        SettingsFilter settingsFilter,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        Supplier<DiscoveryNodes> nodesInCluster
+    ) {
+        if (Version.CURRENT.major == 8) {
+            return List.of(
+                new RestGetActionV7(),
```
Comment (rjernst) [inline modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/RestCompatPlugin.java]: Wouldn't we never have more than one major versions worth of compat in a single version? If that is the case, why do we need to version the class names?

File: modules/rest-compatibility/build.gradle [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+esplugin {
+  description 'Adds a compatiblity layer for the prior major versions REST API'
+  classname 'org.elasticsearch.rest.compat.RestCompatPlugin'
+}
+
+integTest.enabled = false
```
Comment (rjernst) [inline modules/rest-compatibility/build.gradle]: I see tests in this PR; why are they disabled?

File: server/src/main/java/org/elasticsearch/action/DocWriteResponse.java [hunk 26]
```diff
@@ -276,6 +278,9 @@ public void writeTo(StreamOutput out) throws IOException {
     @Override
     public final XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject();
+        if (builder.getCompatibleMajorVersion() == Version.V_7_0_0.major) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/action/DocWriteResponse.java]: If we are just looking at major version number, using the Version constant just to extract "7" seems overkill and less readable than just using the literal `7`.

File: server/src/main/java/org/elasticsearch/rest/RestController.java [hunk 27]
```diff
@@ -340,6 +349,11 @@ private void tryAllHandlers(final RestRequest request, final RestChannel channel
         handleBadRequest(uri, requestMethod, channel);
     }
 
+    public static boolean isV7Compatible(ToXContent.Params params) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/rest/RestController.java]: Why is this V7 specific? Aren't the compatible params version agnostic?

**Review by rjernst — COMMENTED** [review]
This is not an exhaustive review, but I have a few minor comments.

File: modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/RestCompatPlugin.java (hunk 24 above)
Comment (pgomulka) [reply modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/RestCompatPlugin.java]: There is a narrow window where 2 versions can be present.
Let's say we released version 8 (or a feature freeze) and we are now already developing ES9 and there will be features compatible with v8. 
We won't be able to remove the v7 compat features instantly so for some time there will be both.

File: modules/rest-compatibility/build.gradle (hunk 25 above)
Comment (pgomulka) [reply modules/rest-compatibility/build.gradle]: they should be enabled. fixed

File: server/src/main/java/org/elasticsearch/action/DocWriteResponse.java (hunk 26 above)
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/action/DocWriteResponse.java]: I think this will help the removal of compatible code. We have not decided on what mechanism to use in order to help finding this code. 
The plan at the moment is to put as much code v7-compatible into `module/rest-compatibility` 
Using Version.V_7_0_0.major would make sure that if the version is removed after major upgrade it will also highlight the places where v7-compatible  code has to be removed

File: server/src/main/java/org/elasticsearch/rest/RestController.java (hunk 27 above)
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/rest/RestController.java]: good point, fixed

File: modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/RestCompatPlugin.java (hunk 24 above)
Comment (jakelandis) [reply modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/RestCompatPlugin.java]: I am +1 on using the version in the class for this reason, and for the obviousness they provide. While (arguably) ugly, these classes are special in that they are not intended to be maintained, rather they are intended to be deleted after a some point in time.  Having the version in the name makes is more obvious and helps to prevent accidental modifications (like from a simple class search) and makes it really obvious if these things exist for too long.

File: server/src/main/java/org/elasticsearch/action/DocWriteResponse.java (hunk 26 above)
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/action/DocWriteResponse.java]: Also, the removal of the of the Version.V_7_0_0 constant in the code (for v9) will cause this to fail compilation. Which is likely the correct the behavior. It does indeed make major version bumps harder (one more thing that breaks), but for now I think this is the best we can do. It feels like there is something that can be done to ease that upgrade pain, but that is outside the scope of this PR.

File: modules/rest-compatibility/build.gradle (hunk 25 above)
Comment (pgomulka) [reply modules/rest-compatibility/build.gradle]: actually there are no integTests, but there are unit tests so integTests.enabled = false tests.enabled=true

File: server/src/test/java/org/elasticsearch/rest/RestControllerTests.java (hunk 16 above)
Comment (pgomulka) [reply server/src/test/java/org/elasticsearch/rest/RestControllerTests.java]: actually in order to test a compatible handler, a Version.Current - 1 has to be used

**Comment by jakelandis** [comment]
@pgomulka - looking good. I have question about error handling and the combinations of `Accept` and `Content-Type` headers.

Indexing with types and compatibility mode requires both `Accept` and `Content-Type` headers. This is how it is implemented and I believe is the correct behavior. We don't want to open the door (at least not now) for mixing normal requests with compatible responses (or vice-versa).  

For example this works just fine:
```
curl -v -H "Accept: application/vnd.elasticsearch+json;compatible-with=7" -H "Content-Type: application/vnd.elasticsearch+json;compatible-with=7" -X PUT localhost:9200/foo/blah/1 -d '{ "b" : true }'
```

However, I think we could be more helpful for when it appears compatility mode is request. For example the following (correctly) doesn't work :  
```
curl -v  -H "Content-Type: application/vnd.elasticsearch+json;compatible-with=7" -X PUT localhost:9200/foo/blah/1 -d '{ "b" : true }'

curl -v -H "Accept: application/json" -H "Content-Type: application/vnd.elasticsearch+json;compatible-with=7" -X PUT localhost:9200/foo/blah/1 -d '{ "b" : true }'

curl -v -H "Accept: application/vnd.elasticsearch+json;compatible-with=7"  -X PUT localhost:9200/foo/blah/1 -d '{ "b" : true }'
```

I think the above cases are reasonable attempts to trigger compatibility mode, but for (arguably pedantic) reasons don't. I think these should error in a meaningful way, maybe a 406 with a description stating that you need to set both the `Accept` and `Content-Type` header to trigger compatibility when a there is request body ?

I think the logic would be be if no body required on the request then only `Accept` is needed (we can ignore `Content-Type`), if a body is needed, then both `Accept` and `Content-Type` must be sent and equal to each other's compatible request, if a body is needed but only 1 of the `Accept` or `Content-Type` header is sent (or both are sent and they mismatch), then error with a 406 and meaningful error message.  

To put it another way if the `Content-Type` is needed, then so is the `Accept` header, and they need to match w.r.t compatibility, but the `Accept` header by itself is OK assuming there is no body sent in the request. 

Also we may want to surface a warning in the logs around poorly requested no-body requests. For example (the `Content-Type` is being ignored and is also not supported)
```
 curl  -v -H "Accept:: application/vnd.elasticsearch+json;compatible-with=99" -X GET localhost:9200/foo/mytype/1

 curl  -v -H "Accept:: application/vnd.elasticsearch+json;compatible-with=7" -H "Content-Type: application/vnd.elasticsearch+json;compatible-with=99"  -X GET localhost:9200/foo/mytype/1
 ```

Thoughts ?

Also, I am fine if this happens in a follow up PR.

**Comment by pgomulka** [comment]
@jakelandis totally agree with all points raised. I wonder if we should not do this in a separate PR. Will raise it shortly.

**Comment by pgomulka** [comment]
ok to test

**Review by jakelandis — APPROVED** [review]
LGTM. Let's get this in the feature branch as the foundation for other compatibility changes.

**Review by jtibshirani — COMMENTED** [review]
I didn't do an exhaustive review so I won't give an explicit approval, but all my earlier comments have been addressed. Thanks!
